examples/pod_api.rs

```diff
@@ -41,8 +41,8 @@ async fn main() -> anyhow::Result<()> {
             // wait for it..
             std::thread::sleep(std::time::Duration::from_millis(5_000));
         }
-        Err(kube::Error::Api(ae)) => assert_eq!(ae.code, 409), // if you skipped delete, for instance
-        Err(e) => return Err(e.into()),                        // any other case is probably bad
+        Err(kube::Error::Api(ae)) => assert_eq!(ae.code, Some(409)), // if you skipped delete, for instance
```

Comment:
a bit uglier, and a breaking change, but a small one.

Reply:
Yeah, API error handling already felt awkward when writing controllers. Matches like `kube::Error::Api(kube::error::ErrorResponse { code: 404, .. })` are common.

I was going to propose switching to `snafu` and add new `Error` specific for client/API (let's say `kube::client::Error` for now). Providing specific variants like `Error::NotFound` and `Error::Conflict` makes sense because these are very common. We can include our wrapper for `meta::v1::Status` in variants for those who need them.

Also, I'm not sure if there's any reason to keep the deprecated runtime, but I think it makes sense to move `kube-runtime` under `kube` soon and this (`snafu`) is a necessary step.

Reply:
Yeah, there's definitely bigger improvements to be made in client error handling, separating it out more like how it's done for `ConfigError`, and special purpose creating specific errors would help a lot like you say.

A small question though. How are you finding working with `snafu`? What benefits do you find with it over `thiserror`?

I ask because I was kind of thinking to propose unifying under `thiserror` instead because every time i make another app, I have started reaching for it instead because of ergonomics:

- snafu's error enum ends up being 5 lines per impl (when using a source + backtrace) rather than thiserror's 2
- snafu's error variants all have to be imported into scope and causes extra cycles when developing

Reply:
> A small question though. How are you finding working with snafu? What benefits do you find with it over thiserror?

IMO, the big benefit is that the context selectors push you towards having many smaller and more specific error types that act as a sort of semantic backtrace ("file not found while finding Y while reconciling X"), rather than the more or less unmanageable `kube::Error` ("something has gone wrong in library X, have fun figuring out why").

> - snafu's error enum ends up being 5 lines per impl (when using a source + backtrace) rather than thiserror's 2

`snafu` is agnostic to struct vs tuple variants, you can do:

```
#[derive(Debug, Snafu)
pub enum Error {
    #[snafu(display("IO error"))]
    Io(#[snafu(source)] std::io::Error, #[snafu(backtrace)] Backtrace)
}
```

The current state in `kube-derive` is more about my own bias towards struct-style enums.

Backtraces can also be inherited from other `snafu` errors, which would save you an item.

> - snafu's error variants all have to be imported into scope and causes extra cycles when developing

Not really, if you follow `snafu`'s recommended error-enum-per-module style. The enum is the external interface of the module, the context selectors are an implementation detail.

Reply:
Can we move to https://github.com/clux/kube-rs/discussions/453? It's easier for others to join, and we can leave this PR alone.

Reply:
> Providing specific variants like Error::NotFound and Error::Conflict makes sense because these are very common. We can include our wrapper for meta::v1::Status in variants for those who need them.

I agree that that would be more comfortable, but it's a big compatibility footgun, sadly.

Adding a `Conflict` arm would mean that a match on `Err(kube::Error::Api(kube::error::ErrorResponse { code: 409, .. }))` would still compile just fine, but silently never match anything.

Reply:
If we keep `kube::client::Status`, we can add something like:
```rust
impl Status {
    pub fn classify(&self) -> Reason;
}

#[non_exhaustive]
pub enum Reason {
    NotExists,
    Confict,
    PermissionDenied,
    Unknown
}
```
Otherwise, it can be implemented as a free function or an extension trait for `k8s_openapi::Status`